compiler/noirc_evaluator/src/ssa/opt/remove_truncate_after_range_check.rs

```diff
@@ -0,0 +1,138 @@
+use im::HashMap;
+
+use crate::ssa::{
+    ir::{function::Function, instruction::Instruction, value::ValueId},
+    ssa_gen::Ssa,
+};
+
+impl Ssa {
+    /// This SSA pass removes `truncate` instructions that happen on values that
+    /// have a `range_check` on them, where the checked range is less or equal than
+    /// the bits to truncate (the truncate isn't needed then as it won't change the
+    /// underlying value).
+    pub(crate) fn remove_truncate_after_range_check(mut self) -> Self {
+        for function in self.functions.values_mut() {
+            function.remove_truncate_after_range_check();
+        }
+        self
+    }
+}
+
+impl Function {
+    pub(crate) fn remove_truncate_after_range_check(&mut self) {
+        for block in self.reachable_blocks() {
+            // Keeps the minimum bit size a value was range-checked against
+            let mut range_checks: HashMap<ValueId, u32> = HashMap::new();
+
+            let instructions = self.dfg[block].take_instructions();
+            for instruction_id in instructions {
+                let instruction = &self.dfg[instruction_id];
+
+                // If this is a range_check instruction, associate the max bit size with the value
+                if let Instruction::RangeCheck { value, max_bit_size, .. } = instruction {
+                    range_checks
+                        .entry(*value)
+                        .and_modify(|current_max| {
+                            if *max_bit_size < *current_max {
+                                *current_max = *max_bit_size;
+                            }
+                        })
+                        .or_insert(*max_bit_size);
+                }
+
+                // If this is a truncate instruction, check if there's a range check for that same value
+                if let Instruction::Truncate { value, bit_size, .. } = instruction {
+                    if let Some(range_check_bit_size) = range_checks.get(value) {
+                        if range_check_bit_size <= bit_size {
+                            // We need to replace the truncated value with the original one. That is, in:
+                            //
+                            // range_check v0 to 32 bits
+                            // v1 = truncate v0 to 32 bits, max_bit_size: 254
+                            //
+                            // we need to remove the `truncate` and all references to `v1` should now be `v0`.
+                            let result =
+                                self.dfg.instruction_results(instruction_id).first().unwrap();
+                            self.dfg.set_value_from_id(*result, *value);
+                            continue;
+                        }
+                    }
+                }
+
+                self.dfg[block].insert_instruction(instruction_id);
```

Comment:
It would probably be best to use `retain` here to avoid repeated allocations for the `Vec`. :shrug: